Add options for alternate ssh port and reverse tunnel

jasonphillips opened the pull request:
I've found that using the SSH connection option is limited in my use cases when connecting to a server over VPN, or in any situation where (A) I need to use an alternate SSH port, or (B) my local machine cannot be accessed directly by the remote machine.

I added a couple of options to the ssh configuration to help with both of these cases:

 - **port** - an alternative remote ssh port
 - **tunnel** - this solves a problem that was breaking basic pandas <-> SAS functionality when a firewall or security configuration blocks the remote server from accessing arbitrary ports on the local machine; `sasdata2dataframe` needs to open a local port for streaming data in from the SAS session, but the remote machine could not reach it. If you pass a `tunnel` port number, it will now establish a reverse tunnel on that port, using it to transmit the data by default. 

I've written configuration tests that verify these options are being processed correctly without needing to have a live connection that supports them. Note that checking the commands carefully in tests required moving the construction of execv parameters into a separate method (`_buildcommand`). Thoughts welcome!

tomweber-sas commented:
Hey, this sounds cool, thanks! I'm on vacation but will be in next week and will talk a look and see about getting this in.
Thanks!
Tom

tomweber-sas commented:
@jasonphillips , I'm back in the office today and have had a chance to look at this some (more tomorrow). From just a code review, this looks good. I'm trying to catch up w/ a number of things, so I will continue with this tomorrow. But, it looks go to me so far. Just fyi.
Thanks,
Tom

tomweber-sas commented:
@jasonphillips  , I’ve had more time to look at this now. It looks good and thanks for working this out. I do have a few things I’d like to address though.

1) in SASconfigSTDIO, you added the two options initializations (lines 53,54). Can you also add the corresponding blocks of code you’ll find below that which allow for these options to be specified and parsed on the SASsession() method call. You can cut-n-paste the one for ‘ssh’ or ‘host’ and have them for ‘tunnel’ and ‘port’. And, not to impose my OCD on others, but for integer parms, I happen to initialize to None as opposed to a null string ‘’. It all works either way, but was just my convention for int parms as opposed to string parms. See ‘iomport’ and ‘timeout’ in SASconfigIOM. Either way if fine though.

2) adding SO_REUSEADDR to the socket in sasdata2dataframe. Was this necessary for any reason, or just a convention you prefer? I had 2 concerns about this. First if it is actually available and consistent on all of the OS’s. It appears to be consistent enough except for on windows (from what I’ve found). As STDIO isn’t available on windows, that behavioral inconsistency is not problematic. I assume you have this set to get around the SO_LINGER delay when trying to use that specific port repeatedly. However, this also opens up the possibility of other connections attaching to this socket, though we don’t accept but the first connection, so there shouldn’t be any actual exposure to that.

3) adding ‘localhost’ to bind() instead of the current INADDR_ANY (‘’). Won’t this keep a remote connection from connecting? Or is it that ‘localhost’ is needed when using the tunneling option for that to work right and this should only be set when tunneling? I haven’t tried it with tunneling, but when I try this for the usual case I get the following error:
```
ConnectionRefusedError: [Errno 111] Connection refused
```

If my assumptions about 2 and 3 are correct, can you restrict those only to the case when tunneling? I don’t think the normal case will work with 3, as the socket will be bound to 127.0.0.1 only, and a remote connection won’t be able to connect to that. 

If any of my assumptions aren’t right, can you identify the goals of 2 and 3. I haven’t fully tried all of this out yet, some assumptions are just based upon reading the manuals on this, which aren’t the most clear on these two issues. I may be missing something  

Thanks! 
Tom

jasonphillips commented:
Thanks, I've updated in response to your helpful notes:

 1. I've now adjusted so that the port and tunnel options can be provided as arguments, as suggested (and initialized as `None`, indeed a much better convention than the integer / string confusion) and added tests to ensure these arguments are applied / not applied correctly according to lockdown config. I did, however, make the behavior slightly different from the other lock options, which I hope makes sense:  the other arguments are only blocked if lockdown is `True` _and_ the config file already contains a value for the option. Where port and tunnel are concerned, it seemed safer for lockdown mode to also prevent setting these if not set at all in the base config.

 2. Originally I was having some difficulty with testing until I set the SO_REUSEADDR flag, but I've removed it and resumed testing just fine now; it may have just been a deeper collision of sessions at play which doesn't need to be handled. Particularly if that flag leads to any inconsistent platform behavior, I'm fine with taking it back out.

 3. This one was on oversight -- I should indeed have only set 'localhost' as the bind when using the tunnel, where it makes sense given the nature of the reverse tunnel, but should not be used otherwise. I corrected that now with a simple conditional.

tomweber-sas commented:
That all looks good, thanks!
1) looks good, and that's interesting about not letting anyone specify these even if they weren't in the config if lockdown. I think that does make sense here. I might review all the others to see if any others make more sense that way.
2) I'm not convinced yet that we wouldn't need SO_REUSEADDR set for this case (tunneling for sd2df), as we are using a dedicated port number. and consecutive uses may fail due to the SO_LINGER value with the local socket in a TIME_WAIT or CLOSE_WAIT state, But, it's trivial to add in after spending some time playing with it and seeing if it really matters in any situation. I think it's fine, and if we end up needing it we can just add it in.
3) Good. Then I think I do understand it and that seems right.

Ok, for 2), I've played around with this and I see why it isn't needing SO_RESUEADDR. the port on the local side (python), isn't the port specified on the -R parm. It's an open ephemeral port. So there's no confilct requiring addr reuse. The connections do linger, but as the local port varies each connection, there's no conflict. I'm kinda surprised not to see 32701 on both sides???
```
sastpw   14170 14166  0 13:02 pts/5    00:00:00 /usr/bin/ssh -t tom64-2 -p 22 -R 32701:localhost:32701 ...

>>> x = cars.to_df()
(<socket.socket fd=8, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 32701), raddr=('127.0.0.1', 40376)>, ('127.0.0.1', 40376))
>>> y = cars.to_df()
(<socket.socket fd=8, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 32701), raddr=('127.0.0.1', 40378)>, ('127.0.0.1', 40378))
>>> z = cars.to_df()
(<socket.socket fd=8, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 32701), raddr=('127.0.0.1', 40380)>, ('127.0.0.1', 40380))
>>>

tom64-3> netstat -ta | grep 32701
tcp        0      0 localhost:40376             localhost:32701             TIME_WAIT
tcp        0      0 localhost:40380             localhost:32701             TIME_WAIT
tcp        0      0 localhost:40378             localhost:32701             TIME_WAIT

```

So, it appears that the tunneling isn't using 32701 on the client side. I haven't messed w/ ssh -R before, but I'll see what else I can learn. I think this is good as is, cuz we can always tweak it if need be. It's working like a champ for me this way.

Great job! and thanks again!
Tom
